src/redis-cli.c

```diff
@@ -9820,6 +10013,571 @@ void testHintSuite(char *filename) {
+    sendReadOnly();
+
+    /* SCAN loop */
+    do {
+        /* Grab some keys and point to the keys array */
+        reply = sendScan(&it);
+        scan_loops++;
+        keys = reply->element[1];
+
+        /* Reallocate our type and size array if we need to */
+        if (keys->elements > array_size) {
+            memkeys_types = zrealloc(memkeys_types, sizeof(typeinfo*)*keys->elements);
+            memkeys_sizes = zrealloc(memkeys_sizes, sizeof(unsigned long long)*keys->elements);
+
+            bigkeys_types = zrealloc(bigkeys_types, sizeof(typeinfo*)*keys->elements);
+            bigkeys_sizes = zrealloc(bigkeys_sizes, sizeof(unsigned long long)*keys->elements);
+
+            if (!memkeys_types || !memkeys_sizes || !bigkeys_types || !bigkeys_sizes) {
+                fprintf(stderr, "Failed to allocate storage for keys!\n");
+                exit(1);
+            }
+
+            array_size = keys->elements;
+        }
+
+        /* Retrieve types and sizes for memkeys */
+        getKeyTypes(memkeys_types_dict, keys, memkeys_types);
+        getKeySizes(keys, memkeys_types, memkeys_sizes, 1, memkeys_samples);
+
+        /* Retrieve types and sizes for bigkeys */
+        getKeyTypes(bigkeys_types_dict, keys, bigkeys_types);
+        getKeySizes(keys, bigkeys_types, bigkeys_sizes, 0, memkeys_samples);
+
+        for(i=0;i<keys->elements;i++) {
+            /* Skip keys that disappeared between SCAN and TYPE */
+            if (!memkeys_types[i] || !bigkeys_types[i]) {
+                continue;
+            }
+
+            total_size += memkeys_sizes[i];
+            sampled++;
+
+            updateTopSizes(keys->element[i]->str, keys->element[i]->len, memkeys_sizes[i],
+                           memkeys_types[i]->name, top_sizes, top_sizes_limit);
+            updateKeyType(keys->element[i], memkeys_sizes[i], memkeys_types[i]);
+            updateKeyType(keys->element[i], bigkeys_sizes[i], bigkeys_types[i]);
+
+            /* Key Size distribution */
+            if (hdr_record_value(keysize_histogram, memkeys_sizes[i]) == 0) {
+                fprintf(stderr, "Value %llu not added in the hdr histogram.\n", memkeys_sizes[i]);
+            }
+
+            /* Key length distribution */
+            addSizeDist(&key_length_dist, keys->element[i]->len);
+        }
+
+        /* refresh keystats info on regular basis */
+        if (scan_loops % 10 == 0 && (isatty(STDOUT_FILENO) || getenv("FAKETTY"))) {
+             displayKeyStats(sampled, total_keys, total_size, memkeys_types_dict, bigkeys_types_dict,
+                top_sizes, top_sizes_limit, 1);
+        }
+
+        /* Sleep if we've been directed to do so */
+        if (config.interval && (scan_loops % 100) == 0) {
+            usleep(config.interval);
+        }
+
+        freeReplyObject(reply);
+    } while(force_cancel_loop == 0 && it != 0);
+
+    displayKeyStats(sampled, total_keys, total_size, memkeys_types_dict, bigkeys_types_dict, top_sizes,
```

Comment:
what about make the progress to 100% if it is 0(the scan was completed).

```suggestion
    displayKeyStats(it ? sampled : total_keys, total_keys, total_size, memkeys_types_dict, bigkeys_types_dict, top_sizes,
```

Reply:
We could finish the scan with less than 100% if we use `--cursor` (we get the cursor position when we interrupt the scan with Ctrl-C).
When we start the scan at a specific cursor, it might be better to show the user that the information is not on 100% of the keys.

Reply:
i'm okay with both.